text/0102-offchain-parachain-runtime-upgrades.md

```diff
@@ -0,0 +1,354 @@
+
+### Introduce a new UMP message type `RequestCodeUpgrade`
+
+As part of elastic scaling we are already planning to increase flexibility of [UMP
+messages](https://github.com/polkadot-fellows/RFCs/issues/92#issuecomment-2144538974), we can now use this to our advantage and introduce another UMP message:
+
+```rust
+enum UMPSignal {
+  // For elastic scaling
+  OnCore(CoreIndex),
+  // For off-chain upgrades
+  RequestCodeUpgrade(Hash),
+}
+```
+
+We could also make that new message a regular XCM, calling an extrinsic on the
+relay chain, but we will want to look into that message right after validation
+on the backers on the node side, making a straight forward semantic message more
+apt for the purpose.
+
+
+### Handle `RequestCodeUpgrade` on backers
+
+We will introduce a new request/response protocol for both collators and
+validators, with the following request/response:
+
+```rust
+struct RequestBlob {
+  blob_hash: Hash,
+}
+
+struct BlobResponse {
+  blob: Vec<u8>
+}
+```
+
+This protocol will be used by backers to request the PVF from collators in the
+following conditions:
+
+1. They received a collation sending `RequestCodeUpgrade`.
+2. They received a collation, but they don't yet have the code that was
+   previously registered on the relaychain. (E.g. disk pruned, new validator)
+
+In case they received the collation via PoV distribution instead of from the
+collator itself, they will use the exact same message to fetch from the valiator
+they got the PoV from.
+
+### Get the new code to all validators
+
+Once the candidate issuing `RequestCodeUpgrade` got backed on chain, validators
+will start fetching the code from the backers as part of availability
+distribution.
+
+To mitigate attack vectors we should make sure that serving requests for code
+can be treated as low priority requests. Thus I am suggesting the following
+scheme:
+
+Validators will notice via a runtime API (TODO: Define) that a new code has been requested, the
+API will return the `Hash` and a counter, which starts at some configurable
+value e.g. 10. The validators are now aware of the new hash and start fetching,
+but they don't have to wait for the fetch to succeed to sign their bitfield.
+
+Then on each further candidate from that chain that counter gets decremented.
+Validators which have not yet succeeded fetching will now try again. This game
+continues until the counter reached `0`. Now it is mandatory to have to code in
+order to sign a `1` in the bitfield.
+
+PVF pre-checking will happen after the candidate which brought the counter to
+`0` has been successfully included and thus is also able to assume that 2/3 of
+the validators have the code.
+
+This scheme serves two purposes:
+
+1. Fetching can happen over a longer period of time with low priority. E.g. if
+   we waited for the PVF at the very first avaialbility distribution, this might
+   actually affect liveness of other chains on the same core. Distributing
```

Comment:
That's why we have a configurable amount of parachain blocks to do the fetching. If we ever run into availability problems we can:

1. Increase that amount of blocks we have time to fetch the PVF.
2. Limit the amount of runtime upgrades we are willing to do in a timespan and add priority fees (already planned) to requests, to secure a spot in case of competition.

Note however that right now we do distribute those upgrades within a single relaychain slot twice, once via statement distribution then via the relay chain block. In the new scheme if we set the number of required parachain blocks to 10, we reduced pressure 20 times. Thus I doubt it will be a problem in practice and if it ever were, we have means to fix it.